File: transfers/obligations.py
"""
Obligation event emitter.

Converts ObligationEvents from the product portfolio into
concrete Transactions. This is the bridge between the pure
product layer (no infra, no fraud flags) and the transaction
pipeline (with device/IP routing and balance constraints).

Installment-loan behavior modeled here:
- occasional late payments
- missed monthly payments
- partial payments
- catch-up / cure payments
- delinquency clustering once an account falls behind

Why this lives here:
The product layer defines contractual due events.
This transfer layer decides how those due events appear in
observed transaction history.
"""

from dataclasses import dataclass
from datetime import datetime, timedelta

from common.random import Rng
from common.transactions import Transaction
from entities.products.auto_loan import AutoLoanTerms
from entities.products.event import Direction, ObligationEvent
from entities.products.mortgage import MortgageTerms
from entities.products.portfolio import PortfolioRegistry
from entities.products.student_loan import StudentLoanTerms
from transfers.factory import TransactionDraft, TransactionFactory

type InstallmentTerms = MortgageTerms | AutoLoanTerms | StudentLoanTerms


@dataclass(slots=True)
class _InstallmentState:
    """Running delinquency state for one person/product pair."""

    carry_due: float = 0.0
    delinquent_cycles: int = 0


_LOAN_PRODUCT_TYPES: frozenset[str] = frozenset(
    {"mortgage", "auto_loan", "student_loan"}
)


def _round_money(amount: float) -> float:
    return round(max(0.0, float(amount)), 2)


def _add_basic_jitter(rng: Rng, ts: datetime) -> datetime:
    """
    Add small timing jitter to scheduled non-loan events.

    Real payments do not land at exactly midnight on the due date.
    ACH processing, autopay timing, and manual habits introduce
    0-3 days and variable hours of offset.
    """
    return ts + timedelta(
        days=rng.int(0, 3),
        hours=rng.int(0, 12),
        minutes=rng.int(0, 60),
    )


def _uniform_between(rng: Rng, low: float, high: float) -> float:
    """
    Uniform float in [low, high].

    Uses Rng.float() because the wrapper does not expose a dedicated
    uniform() helper.
    """
    low_f = float(low)
    high_f = float(high)

    if high_f < low_f:
        raise ValueError("high must be >= low")

    if high_f == low_f:
        return low_f

    u = rng.float()
    return low_f + ((high_f - low_f) * u)


def _payment_timestamp(
    rng: Rng,
    due_ts: datetime,
    *,
    late_p: float,
    late_days_min: int,
    late_days_max: int,
    force_late: bool = False,
) -> datetime:
    """
    Timestamp for a realized payment.

    Late payments are shifted by the configured late-day range.
    On-time payments still get hour/minute jitter within the day.
    """
    should_be_late = force_late or rng.coin(late_p)

    if should_be_late:
        delay_days = 0
        if late_days_max > 0:
            delay_days = rng.int(late_days_min, late_days_max + 1)

        return due_ts + timedelta(
            days=delay_days,
            hours=rng.int(8, 22),
            minutes=rng.int(0, 60),
        )

    return due_ts + timedelta(
        hours=rng.int(6, 18),
        minutes=rng.int(0, 60),
    )


def _make_txn(
    *,
    txf: TransactionFactory,
    src: str,
    dst: str,
    amount: float,
    timestamp: datetime,
    channel: str,
) -> Transaction:
    return txf.make(
        TransactionDraft(
            source=src,
            destination=dst,
            amount=_round_money(amount),
            timestamp=timestamp,
            channel=channel,
        )
    )


def _resolve_terms(
    registry: PortfolioRegistry,
    event: ObligationEvent,
) -> InstallmentTerms | None:
    """
    Resolve the concrete installment-loan terms backing this obligation event.
    """
    portfolio = registry.get(event.person_id)
    if portfolio is None:
        return None

    if event.product_type == "mortgage":
        return portfolio.mortgage
    if event.product_type == "auto_loan":
        return portfolio.auto_loan
    if event.product_type == "student_loan":
        return portfolio.student_loan

    return None


def _emit_plain_event(
    *,
    rng: Rng,
    txf: TransactionFactory,
    event: ObligationEvent,
    person_acct: str,
    end_excl: datetime,
) -> Transaction | None:
    """
    Emit a non-installment obligation with light timestamp jitter only.
    """
    ts = _add_basic_jitter(rng, event.timestamp)
    if ts >= end_excl:
        return None

    if event.direction == Direction.OUTFLOW:
        src = person_acct
        dst = event.counterparty_acct
    else:
        src = event.counterparty_acct
        dst = person_acct

    return _make_txn(
        txf=txf,
        src=src,
        dst=dst,
        amount=event.amount,
        timestamp=ts,
        channel=event.channel,
    )


def _effective_prob(base: float, multiplier: float) -> float:
    return min(0.95, max(0.0, base * multiplier))


def _installment_payment_amount(
    *,
    rng: Rng,
    total_due: float,
    partial_min_frac: float,
    partial_max_frac: float,
) -> float:
    """
    Realized amount for a partial installment payment.
    """
    frac = _uniform_between(rng, partial_min_frac, partial_max_frac)
    raw_paid = total_due * frac
    paid = min(total_due, max(0.01, raw_paid))
    return _round_money(paid)


def _emit_installment_event(
    *,
    registry: PortfolioRegistry,
    state_by_key: dict[tuple[str, str], _InstallmentState],
    rng: Rng,
    txf: TransactionFactory,
    event: ObligationEvent,
    person_acct: str,
    end_excl: datetime,
) -> list[Transaction]:
    """
    Emit one scheduled installment due event as zero or one observed
    transaction depending on delinquency behavior.

    Current modeled cases:
    - missed payment -> no txn, amount carried forward
    - partial payment -> one txn for a fraction of total due
    - cure payment -> one txn that catches up all arrears
    - normal payment -> one full scheduled payment
    """
    terms = _resolve_terms(registry, event)
    if terms is None:
        txn = _emit_plain_event(
            rng=rng,
            txf=txf,
            event=event,
            person_acct=person_acct,
            end_excl=end_excl,
        )
        return [] if txn is None else [txn]

    state_key = (event.person_id, event.product_type)
    state = state_by_key.setdefault(state_key, _InstallmentState())

    scheduled_amount = _round_money(event.amount)
    total_due = _round_money(scheduled_amount + state.carry_due)

    if total_due <= 0.0:
        return []

    late_p = terms.late_p
    late_days_min = terms.late_days_min
    late_days_max = terms.late_days_max
    miss_p = terms.miss_p
    partial_p = terms.partial_p
    cure_p = terms.cure_p
    cluster_mult = terms.cluster_mult
    partial_min_frac = terms.partial_min_frac
    partial_max_frac = terms.partial_max_frac

    delinquency_mult = 1.0
    if state.delinquent_cycles > 0:
        delinquency_mult = cluster_mult ** min(state.delinquent_cycles, 2)

    effective_late_p = _effective_prob(late_p, delinquency_mult)
    effective_miss_p = _effective_prob(miss_p, delinquency_mult)
    effective_partial_p = _effective_prob(
        partial_p,
        1.0 + (0.5 * state.delinquent_cycles),
    )

    txns: list[Transaction] = []

    # Catch-up / cure payment for previously unpaid balance.
    if state.carry_due > 0.0:
        cure_boost = 1.0 + (0.25 * state.delinquent_cycles)
        effective_cure_p = min(0.98, cure_p * cure_boost)

        if rng.coin(effective_cure_p):
            cure_ts = _payment_timestamp(
                rng,
                event.timestamp,
                late_p=max(effective_late_p, 0.75),
                late_days_min=late_days_min,
                late_days_max=late_days_max,
                force_late=True,
            )
            if cure_ts < end_excl:
                txns.append(
                    _make_txn(
                        txf=txf,
                        src=person_acct,
                        dst=event.counterparty_acct,
                        amount=total_due,
                        timestamp=cure_ts,
                        channel=event.channel,
                    )
                )
                state.carry_due = 0.0
                state.delinquent_cycles = 0

            return txns

    decision_u = rng.float()

    # Missed payment: nothing posts this cycle; full due carries forward.
    if decision_u < effective_miss_p:
        state.carry_due = total_due
        state.delinquent_cycles += 1
        return []

    decision_u -= effective_miss_p

    # Partial payment: some amount posts, remainder carries forward.
    if decision_u < effective_partial_p:
        paid_amount = _installment_payment_amount(
            rng=rng,
            total_due=total_due,
            partial_min_frac=partial_min_frac,
            partial_max_frac=partial_max_frac,
        )
        unpaid_amount = _round_money(total_due - paid_amount)

        if paid_amount > 0.0:
            partial_ts = _payment_timestamp(
                rng,
                event.timestamp,
                late_p=effective_late_p,
                late_days_min=late_days_min,
                late_days_max=late_days_max,
            )
            if partial_ts < end_excl:
                txns.append(
                    _make_txn(
                        txf=txf,
                        src=person_acct,
                        dst=event.counterparty_acct,
                        amount=paid_amount,
                        timestamp=partial_ts,
                        channel=event.channel,
                    )
                )

        state.carry_due = unpaid_amount
        if unpaid_amount > 0.0:
            state.delinquent_cycles += 1
        else:
            state.delinquent_cycles = 0

        return txns

    # Full scheduled-cycle payment.
    full_ts = _payment_timestamp(
        rng,
        event.timestamp,
        late_p=effective_late_p,
        late_days_min=late_days_min,
        late_days_max=late_days_max,
    )
    if full_ts < end_excl:
        txns.append(
            _make_txn(
                txf=txf,
                src=person_acct,
                dst=event.counterparty_acct,
                amount=scheduled_amount,
                timestamp=full_ts,
                channel=event.channel,
            )
        )

    # If arrears still remain from prior cycles, delinquency persists.
    if state.carry_due > 0.0:
        state.delinquent_cycles += 1
    else:
        state.delinquent_cycles = 0

    return txns


def emit(
    registry: PortfolioRegistry,
    *,
    start: datetime,
    end_excl: datetime,
    primary_accounts: dict[str, str],
    rng: Rng,
    txf: TransactionFactory,
) -> list[Transaction]:
    """
    Convert all scheduled obligation events into Transactions.

    Skips events where the person has no primary account
    (external-only or unresolvable).

    Installment loans get delinquency-aware realization logic.
    Other obligations keep the original light-jitter behavior.
    """
    txns: list[Transaction] = []
    state_by_key: dict[tuple[str, str], _InstallmentState] = {}

    for event in registry.all_events(start, end_excl):
        person_acct = primary_accounts.get(event.person_id)
        if not person_acct:
            continue

        if (
            event.direction == Direction.OUTFLOW
            and event.product_type in _LOAN_PRODUCT_TYPES
        ):
            txns.extend(
                _emit_installment_event(
                    registry=registry,
                    state_by_key=state_by_key,
                    rng=rng,
                    txf=txf,
                    event=event,
                    person_acct=person_acct,
                    end_excl=end_excl,
                )
            )
            continue

        txn = _emit_plain_event(
            rng=rng,
            txf=txf,
            event=event,
            person_acct=person_acct,
            end_excl=end_excl,
        )
        if txn is not None:
            txns.append(txn)

    txns.sort(key=lambda t: t.timestamp)
    return txns

File: src/transfers/obligations.cpp
#include "phantomledger/transfers/obligations.hpp"

#include "phantomledger/transfers/helpers.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <unordered_map>

namespace PhantomLedger::transfers::obligations {
namespace {

using Key = entity::Key;
using PersonId = entity::PersonId;
using TimePoint = time::TimePoint;
using entity::product::Direction;
using entity::product::ObligationEvent;
using entity::product::PortfolioRegistry;
using entity::product::ProductType;

// ----------------------------------------------------------------
// Per-account delinquency state
// ----------------------------------------------------------------

struct InstallmentState {
  /// Unpaid balance carried forward from previous cycles. Zero when
  /// the account is current.
  double carryDue = 0.0;

  /// Consecutive cycles with missing or partial payments. Reset to
  /// zero on a full or cure payment.
  int delinquentCycles = 0;
};

struct StateKey {
  PersonId person{};
  ProductType productType = ProductType::unknown;

  [[nodiscard]] bool operator==(const StateKey &) const noexcept = default;
};

struct StateKeyHash {
  [[nodiscard]] std::size_t operator()(const StateKey &k) const noexcept {
    const auto pid = static_cast<std::uint64_t>(k.person);
    const auto pt = static_cast<std::uint64_t>(k.productType);
    return std::hash<std::uint64_t>{}((pid << 8U) ^ pt);
  }
};

// ----------------------------------------------------------------
// Small helpers
// ----------------------------------------------------------------

/// Uniform double in [0, 1). Composed from `uniformInt` because it is
/// the one Rng primitive this module can rely on without asserting
/// the presence of a dedicated u01 method. Uses all 53 mantissa bits
/// of a double so the distribution covers every representable value
/// in the unit interval.
[[nodiscard]] double uniformUnit(random::Rng &rng) noexcept {
  constexpr std::int64_t kMantissa = 1LL << 53;
  const auto i = rng.uniformInt(0, kMantissa);
  return static_cast<double>(i) / static_cast<double>(kMantissa);
}

/// Uniform float in [low, high]. Matches Python `_uniform_between`.
[[nodiscard]] double uniformBetween(random::Rng &rng, double low, double high) {
  if (high < low) {
    throw std::invalid_argument("uniformBetween: high must be >= low");
  }
  if (high == low) {
    return low;
  }
  return low + (high - low) * uniformUnit(rng);
}

/// Light jitter for non-installment events: 0-2 days + 0-11 hours +
/// 0-59 minutes. Matches Python `_add_basic_jitter` exactly (Python
/// uses rng.int(0, 3) -> {0, 1, 2} and rng.int(0, 12) -> {0..11}).
[[nodiscard]] TimePoint addBasicJitter(random::Rng &rng,
                                       TimePoint ts) noexcept {
  const auto days = rng.uniformInt(0, 3);
  const auto hours = rng.uniformInt(0, 12);
  const auto minutes = rng.uniformInt(0, 60);
  return ts + time::Days{days} + time::Hours{hours} + time::Minutes{minutes};
}

/// Timestamp for a realized installment payment. Late payments get a
/// day offset in [lateDaysMin, lateDaysMax] plus hour/minute jitter;
/// on-time payments still get hour/minute jitter within the due day.
[[nodiscard]] TimePoint paymentTimestamp(random::Rng &rng, TimePoint dueTs,
                                         double lateP, std::int32_t lateDaysMin,
                                         std::int32_t lateDaysMax,
                                         bool forceLate) {
  const bool beLate = forceLate || rng.coin(lateP);

  if (beLate) {
    int delayDays = 0;
    if (lateDaysMax > 0) {
      // Python rng.int(lateDaysMin, lateDaysMax + 1) is [lateDaysMin,
      // lateDaysMax]. C++ uniformInt(low, high) is [low, high), so we
      // pass (lateDaysMin, lateDaysMax + 1).
      delayDays =
          static_cast<int>(rng.uniformInt(lateDaysMin, lateDaysMax + 1));
    }
    const auto hour = rng.uniformInt(8, 22);
    const auto minute = rng.uniformInt(0, 60);
    return dueTs + time::Days{delayDays} + time::Hours{hour} +
           time::Minutes{minute};
  }

  const auto hour = rng.uniformInt(6, 18);
  const auto minute = rng.uniformInt(0, 60);
  return dueTs + time::Hours{hour} + time::Minutes{minute};
}

/// Compose a draft with the shared rounding/timestamp helpers.
[[nodiscard]] transactions::Transaction
makeTxn(const transactions::Factory &txf, const Key &src, const Key &dst,
        double amount, TimePoint ts, channels::Tag channel) {
  transactions::Draft draft;
  draft.source = src;
  draft.destination = dst;
  draft.amount = helpers::roundMoney(amount);
  draft.timestamp = helpers::draftTimestamp(ts);
  draft.channel = channel;
  return txf.make(draft);
}

/// Cap an effective probability at 0.95 to guarantee a non-trivial
/// tail on every Bernoulli trial.
[[nodiscard]] constexpr double effectiveProb(double base,
                                             double multiplier) noexcept {
  const double x = base * multiplier;
  return std::clamp(x, 0.0, 0.95);
}

/// Uniform-fraction partial payment amount.
[[nodiscard]] double installmentPaymentAmount(random::Rng &rng, double totalDue,
                                              double partialMinFrac,
                                              double partialMaxFrac) {
  const double frac = uniformBetween(rng, partialMinFrac, partialMaxFrac);
  const double raw = totalDue * frac;
  const double paid = std::min(totalDue, std::max(0.01, raw));
  return helpers::roundMoney(paid);
}

// ----------------------------------------------------------------
// Plain (non-installment) event emission
// ----------------------------------------------------------------

/// Emit a non-installment event with light timestamp jitter. Returns
/// std::nullopt (via an empty-or-one-element vector push) if the
/// jittered timestamp falls outside the active window.
void emitPlainEvent(random::Rng &rng, const transactions::Factory &txf,
                    const ObligationEvent &event, const Key &personAcct,
                    TimePoint endExcl,
                    std::vector<transactions::Transaction> &out) {
  const auto ts = addBasicJitter(rng, event.timestamp);
  if (ts >= endExcl) {
    return;
  }

  Key src{};
  Key dst{};
  if (event.direction == Direction::outflow) {
    src = personAcct;
    dst = event.counterpartyAcct;
  } else {
    src = event.counterpartyAcct;
    dst = personAcct;
  }

  out.push_back(makeTxn(txf, src, dst, event.amount, ts, event.channel));
}

// ----------------------------------------------------------------
// Installment-loan event emission (the state machine)
// ----------------------------------------------------------------

/// Emit one scheduled installment due event as zero or one observed
/// transaction, updating the per-account state in-place.
///
/// Cases:
///   - carry_due > 0 + cure draws -> one cure txn clearing arrears
///   - miss draws                  -> no txn, full due carries forward
///   - partial draws               -> one partial txn; remainder carries
///   - normal                      -> one full scheduled payment
void emitInstallmentEvent(
    const PortfolioRegistry &registry,
    std::unordered_map<StateKey, InstallmentState, StateKeyHash> &stateByKey,
    random::Rng &rng, const transactions::Factory &txf,
    const ObligationEvent &event, const Key &personAcct, TimePoint endExcl,
    std::vector<transactions::Transaction> &out) {

  const auto *terms =
      registry.installmentTerms(event.personId, event.productType);
  if (terms == nullptr) {
    emitPlainEvent(rng, txf, event, personAcct, endExcl, out);
    return;
  }

  const StateKey key{event.personId, event.productType};
  auto &state = stateByKey[key]; // inserts default-constructed state

  const double scheduledAmount = helpers::roundMoney(event.amount);
  const double totalDue = helpers::roundMoney(scheduledAmount + state.carryDue);
  if (totalDue <= 0.0) {
    return;
  }

  // Delinquency multiplier: capped at two applications so the
  // probabilities do not explode once an account has been behind for
  // many cycles. Effectively, a streak of 3+ cycles looks like a
  // streak of 2 for transition purposes.
  double delinquencyMult = 1.0;
  if (state.delinquentCycles > 0) {
    const int exp = std::min(state.delinquentCycles, 2);
    delinquencyMult = std::pow(terms->clusterMult, exp);
  }

  const double effectiveLateP = effectiveProb(terms->lateP, delinquencyMult);
  const double effectiveMissP = effectiveProb(terms->missP, delinquencyMult);
  const double effectivePartialP =
      effectiveProb(terms->partialP,
                    1.0 + (0.5 * static_cast<double>(state.delinquentCycles)));

  // --- Cure branch ---
  //
  // If there's already unpaid balance, test the cure probability
  // first. A successful cure clears arrears and resets delinquency.
  if (state.carryDue > 0.0) {
    const double cureBoost =
        1.0 + (0.25 * static_cast<double>(state.delinquentCycles));
    const double effectiveCureP = std::min(0.98, terms->cureP * cureBoost);

    if (rng.coin(effectiveCureP)) {
      const auto cureTs =
          paymentTimestamp(rng, event.timestamp, std::max(effectiveLateP, 0.75),
                           terms->lateDaysMin, terms->lateDaysMax,
                           /*forceLate=*/true);
      if (cureTs < endExcl) {
        out.push_back(makeTxn(txf, personAcct, event.counterpartyAcct, totalDue,
                              cureTs, event.channel));
        state.carryDue = 0.0;
        state.delinquentCycles = 0;
      }
      return;
    }
  }

  // --- Three-way decision: miss / partial / full ---
  //
  // Draw one u in [0, 1) and walk down the probability ladder. This
  // keeps the three branches mutually exclusive.
  double decisionU = uniformUnit(rng);

  // Missed payment.
  if (decisionU < effectiveMissP) {
    state.carryDue = totalDue;
    state.delinquentCycles += 1;
    return;
  }
  decisionU -= effectiveMissP;

  // Partial payment.
  if (decisionU < effectivePartialP) {
    const double paidAmount = installmentPaymentAmount(
        rng, totalDue, terms->partialMinFrac, terms->partialMaxFrac);
    const double unpaidAmount = helpers::roundMoney(totalDue - paidAmount);

    if (paidAmount > 0.0) {
      const auto partialTs = paymentTimestamp(
          rng, event.timestamp, effectiveLateP, terms->lateDaysMin,
          terms->lateDaysMax, /*forceLate=*/false);
      if (partialTs < endExcl) {
        out.push_back(makeTxn(txf, personAcct, event.counterpartyAcct,
                              paidAmount, partialTs, event.channel));
      }
    }

    state.carryDue = unpaidAmount;
    state.delinquentCycles =
        (unpaidAmount > 0.0) ? state.delinquentCycles + 1 : 0;
    return;
  }

  // Full scheduled payment.
  const auto fullTs =
      paymentTimestamp(rng, event.timestamp, effectiveLateP, terms->lateDaysMin,
                       terms->lateDaysMax, /*forceLate=*/false);
  if (fullTs < endExcl) {
    out.push_back(makeTxn(txf, personAcct, event.counterpartyAcct,
                          scheduledAmount, fullTs, event.channel));
  }

  // If prior carry is still outstanding (scheduled payment only
  // covers the current cycle), delinquency persists. Otherwise, the
  // account is current.
  if (state.carryDue > 0.0) {
    state.delinquentCycles += 1;
  } else {
    state.delinquentCycles = 0;
  }
}

} // namespace

// ----------------------------------------------------------------
// Public API
// ----------------------------------------------------------------

std::vector<transactions::Transaction> emit(const PortfolioRegistry &registry,
                                            TimePoint start, TimePoint endExcl,
                                            const Population &population,
                                            random::Rng &rng,
                                            const transactions::Factory &txf) {
  std::vector<transactions::Transaction> out;

  std::unordered_map<StateKey, InstallmentState, StateKeyHash> stateByKey;

  for (const auto &event : registry.allEvents(start, endExcl)) {
    const auto acctIt = population.primaryAccounts->find(event.personId);
    if (acctIt == population.primaryAccounts->end()) {
      continue;
    }
    const Key personAcct = acctIt->second;

    const bool isInstallmentOutflow =
        event.direction == Direction::outflow &&
        entity::product::isInstallmentLoan(event.productType);

    if (isInstallmentOutflow) {
      emitInstallmentEvent(registry, stateByKey, rng, txf, event, personAcct,
                           endExcl, out);
    } else {
      emitPlainEvent(rng, txf, event, personAcct, endExcl, out);
    }
  }

  // Sort by timestamp for deterministic output. The full audit
  // comparator is unnecessary here because there's no cross-event
  // ordering to break.
  std::sort(out.begin(), out.end(),
            [](const transactions::Transaction &a,
               const transactions::Transaction &b) noexcept {
              return a.timestamp < b.timestamp;
            });

  return out;
}

} // namespace PhantomLedger::transfers::obligations
